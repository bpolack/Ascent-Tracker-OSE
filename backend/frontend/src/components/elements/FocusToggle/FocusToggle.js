import React, { Component } from 'react'
import { connect } from 'react-redux';
import { setFocus } from '../../../actions/focus';
import PropTypes from 'prop-types';

// Other Components
import { Switch } from "@headlessui/react";

export class FocusToggle extends Component {

	constructor(props) {
		super(props);
		this.handleFocus = this.handleFocus.bind(this);
	}

	handleFocus(event) {
		this.props.setFocus(!this.props.focus);
	}

	render() {

		const { focus } = this.props;

		return (
			<div className="bg-ascent-std absolute top-4 left-5 p-3 rounded-full z-20">
				<Switch.Group>
					<div className="flex items-center">
						<Switch.Label className="mr-4 text-ascent-std">Focus Mode</Switch.Label>
						<Switch
							checked={focus}
							onChange={this.handleFocus}
							className={`${focus ? "bg-pink-500" : "bg-gray-700"} 
									relative inline-flex items-center h-6 rounded-full w-11 transition-colors focus:outline-none`}
						>
							<span
								className={`${focus ? "translate-x-6" : "translate-x-1"} 
									inline-block w-4 h-4 transform bg-white rounded-full transition-transform`}
							/>
						</Switch>
					</div>
				</Switch.Group>
			</div>
		)
	}
}

FocusToggle.propTypes = {
	setFocus: PropTypes.func.isRequired,
	focus: PropTypes.bool.isRequired
}

const mapStateToProps = state => ({
	focus: state.focus
});

export default connect(mapStateToProps, { setFocus })(FocusToggle)
