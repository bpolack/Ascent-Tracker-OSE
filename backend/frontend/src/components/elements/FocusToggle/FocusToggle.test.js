import React from 'react';
import { shallow } from 'enzyme';
import { FocusToggle } from './FocusToggle';
import { Switch } from "@headlessui/react";

describe('FocusToggle', () => {
	it("Renders correctly without crashing", () => {
		const setFocusMock = jest.fn();
		
		const wrapper = shallow(<FocusToggle setFocus={setFocusMock} focus={false} />);
		expect(wrapper).toMatchSnapshot();
	});
	it("Renders as toggled with focus prop", () => {
		const setFocusMock = jest.fn();
		
		const wrapper = shallow(<FocusToggle setFocus={setFocusMock} focus={true} />);
		
		const toggleElement = wrapper.find(Switch); 
		expect(toggleElement.prop('checked')).toBe(true);
	});

	describe('When user clicks toggle', () => {
		it('setFocus is triggered with the opposite of focus bool', () => {
			const setFocusMock = jest.fn();

			const wrapper = shallow(<FocusToggle setFocus={setFocusMock} focus={false} />);

			const toggleElement = wrapper.find(Switch); 
			toggleElement.invoke('onChange')();

			expect(setFocusMock).toHaveBeenCalledTimes(1);
			expect(setFocusMock).toHaveBeenCalledWith(true);
		});
	});
});