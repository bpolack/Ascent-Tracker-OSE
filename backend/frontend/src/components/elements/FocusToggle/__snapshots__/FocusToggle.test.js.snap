// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusToggle Renders correctly without crashing 1`] = `ShallowWrapper {}`;
